### PR TITLE
Draft: Add RedirectToActionOther method.

### LINKED ISSUE #36172: Better Support HTTP 303 Status Code
Before you close this as a duplicate, read the post over. I'm aware of the previous request at #13359 but don't feel it adequately addresses the issue.

### Motivation

303 needs to be more easily accessible because its intended use case is *extremely* common:

> It is primarily used to allow the output of a POST action to redirect the user agent to a selected resource, since doing so provides the information corresponding to the POST response in a form that can be separately identified, bookmarked, and cached, independent of the original request.

From [RFC 7231 Hypertext Transfer Protocol (HTTP/1.1): Semantics and Content](https://datatracker.ietf.org/doc/html/rfc7231)

This use case describes almost every endpoint that creates a new persistent resource in an HTML website: you create the resource and then send the user to the page displaying the new resource, typically with a URL containing a unique identifier. In particular, the fact that 303 *must* change the method to a GET request is necessary for these use cases to avoid attempting to create or update the new resource. While 302 *typically* does replace the method in implementation, the standard does not specifically require it the way 303 does. This ambiguity is exactly what I wish to avoid; the client must *not* send another POST request to the new resource's location.

Actually, I wish to avoid 302 entirely because of this ambiguity. It is much better to use a redirect status code with explicit semantics regarding the method than one that lets the client decide what to do. Frameworks should not make it easier to use an ambiguous, less predictable web standard than to use the unambiguous replacements specifically created in response to the ambiguity.

### Status Quo

Currently, the only way to redirect using the 303 status code is to build up the response from scratch. For example, this code might appear in a class extending `Controller`:

```
string location = this.Url.Action("MyAction", new { resourceId = newId });
this.Response.Headers.Add("Location", location);
return this.StatusCode(StatusCodes.Status303SeeOther);
```

This usage pattern is not even well documented.

### Describe the solution you'd like

I'd much rather my code look more like this, similar to what's available for 301, 302, 307, and 308:

```
return this.RedirectRetrieveToAction("MyAction", new { resourceId = newId });
```

I'm not picky about the name. That's just a first stab at it, taken from the wording of RFC specification and the naming pattern already present on `ControllerBase`.

### PR DESCRIPTION
**Pull request by minecraftchest1**
# Add RedirectToActionOther Method.

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [ ] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Summary of the changes (Less than 80 chars)

## Description

Adds the method RedirectToActionOther to the class Microsoft.AspNetCore.Mvc that ahould allow redirecting to an action without preserving the HTTP Request Method.

Fixes #36172

### CODE REVIEW

**Comment by mitchdenny** [comment]
We probably want to take this through API review. Can you update the original issue using this template:

https://github.com/dotnet/aspnetcore/issues/new?assignees=&labels=api-suggestion&projects=&template=30_api_proposal.md&title=

This helps ensure the right people get eyes on it can can tweak things. Your PR build failed due to some code formatting issues and the fact that the public API text file needs to be updated. If you hover over the new method there should be a Fix in VS that updates this file correctly for you.

**Comment by minecraftchest1** [comment]
> We probably want to take this through API review. Can you update the original issue using this template:
> 
> https://github.com/dotnet/aspnetcore/issues/new?assignees=&labels=api-suggestion&projects=&template=30_api_proposal.md&title=
> 
> This helps ensure the right people get eyes on it can can tweak things. Your PR build failed due to some code formatting issues and the fact that the public API text file needs to be updated. If you hover over the new method there should be a Fix in VS that updates this file correctly for you.

I removed the extra newlines from my code. I don't have accessmto Visual Studio (I am runing linux), and VSCode makes my laptop want to fall over dead. Can you point me to file I should update please?

**Comment by minecraftchest1** [comment]
I believe I made the correct changes to the PublicApi file.

File: src/Mvc/Mvc.Core/src/ControllerBase.cs [hunk 1]
```diff
@@ -564,6 +564,123 @@ public virtual RedirectToActionResult RedirectToAction(string? actionName, strin
...
+    /// {
+    ///     var product = RetrieveProduct(id);
+    ///     return View(product);
+    /// }
+    ///
+    /// [HttpPost]
+    /// public IActionResult Product(int id, Product product)
+    /// {
+    ///     UpdateProduct(product);
+    ///     return RedirectToActionOthet();
```
Comment (Eilon) [inline src/Mvc/Mvc.Core/src/ControllerBase.cs:587]: Typo in the sample code: `othet`

**Comment by minecraftchest1** [comment]
Thanks. I will fix it when I get off work. I typed it up late last night, so typo's don't supprise me.



Sent from my T-Mobile 4G LTE Device

File: src/Mvc/Mvc.Core/src/ControllerBase.cs (hunk 1 above)
Comment (minecraftchest1) [reply src/Mvc/Mvc.Core/src/ControllerBase.cs]: Fixed. Pushing change now.

**Comment by minecraftchest1** [comment]
All requested changes made. What else do I need to do?

**Comment by mitchdenny** [comment]
Because this is a change/addition to the public API we need to actually do the API review. When you create a new issue you can select the API proposal template which has all the details that need to be covered. You can either create a new issue or update the original link issue with the contents from the template. Link it to this PR so we can track everything. Then we can look at getting your proposal put through API review.

**Comment by minecraftchest1** [comment]
/azp run



Sent from my T-Mobile 4G LTE Device

**Comment by captainsafia** [comment]
@minecraftchest1 The `/azp run` command is limited to maintainers at the moment. As mentioned in [this comment](https://github.com/dotnet/aspnetcore/pull/48823#issuecomment-1596438528) you'll want to open an API review issue for this PR first to move to the next step. Do you need help with that?

**Comment by minecraftchest1** [comment]
Yes.



Sent from my T-Mobile 4G LTE Device

**Comment by captainsafia** [comment]
You can find the API review template [here](https://github.com/dotnet/aspnetcore/issues/new?assignees=&labels=api-suggestion&projects=&template=30_api_proposal.md&title=). There are descriptions under each heading of the template of the content that's needed.

**Comment by aetos382** [comment]
This implementation does not resolve the request in #36172.
The result of `RedirectToActionOther` is identical to `RedirectToAction`.
The [RedirectToActionResultExecutor](https://github.com/dotnet/aspnetcore/blob/main/src/Mvc/Mvc.Core/src/Infrastructure/RedirectToActionResultExecutor.cs) must be changed to generate a 303 response.

**Comment by aetos382** [comment]
This implementation does not resolve the request in #36172.
The result of RedirectToActionOther is identical to RedirectToAction.

The response generated by `RedirectToActionResultExecutor` is determined by the parameters of `RedirectToActionResult` as follows.

|Method of Controller|preserveMethod|permanent|response|
|-|-|-|-|
|RedirectToAction|false|false|302|
|RedirectToActionPermanent|false|true|301|
|RedirectToActionPreserveMethod|true|false|307|
|RedirectToActionPermanentPreserveMethod|true|true|308|

The problem is that the `preserveMethod` parameter of type `bool` cannot express the semantics that "the request must be made with the GET method".
However, I disagree with the idea of making it an enumerated type representing the three options.
Suppose we create a new type like the following.

```cs
enum RedirectionMethod
{
    AllowChange,
    Preserve,
    MustGet
}
```

In this case, the `permanent` parameter is meaningless for `MustGet`.
For this reason, I think a new Result type and Executor for it should be added to be able to generate a 303 response.

The response codes 301, 302, 307, and 308 all (essentially) mean "The user's request has not been completed and the user agent must retry the request for the given URL."
The 303 response differs from them in that it indicates that "The user's request has been completed and the server is presenting next instruction."
Therefore, I do not believe that 303 should be treated as a variation of other redirects.

**Comment by codlin** [comment]
Is the 303 issue still unresolved?

**Pull request closed without merging**